Experimental : CollectPrimitiveVariables

danieldresser opened the pull request:
John and I already have an ongoing email thread, so some of this may be headed in the wrong direction depending on the outcome of that thread, but I thought I should put up my prototype for reference.

This is a node that allows building some goofy primitive variable processing stuff - it lets you create copies of input primitive variables evaluated in different contexts.  By combining it with a time warp, you can create copies of a primVar evaluated at different times.

In particular, I wanted to demonstrate the performance impact of the slightly weird bool output plug that indicates whether there is any variation in the input in any of the contexts.  This is currently named "isVariation", and it badly needs a better name.  Maybe I should have it mean the reverse ... "allIdentical"?

Anyway, aside from the nuisance of naming it, what this allows is skipping processing of locations where the object is stationary, or only transforming rigidly.  I'll try to paste in my test scene, where I've got one animated sphere, and 8000 static spheres with unique vertex positions.  My toy problem is a network that averages out the current vertex positions with one frame before and after.  I've got two networks that perform this task using CollectPrimitiveVariables, one taking advantage of the "isIdentical" output, and the other not.

`gaffer stats collectPrim.gfr -scene AveragePosition.out`
vs
`gaffer stats collectPrim.gfr -scene AveragePositionWithout.out`
yields 3 seconds vs 6 seconds.

This example is somewhat constructed to show a clear difference, but incredibly vast numbers of heavy non-deforming objects is about as common as it gets in production scenes.  I think this optimisation is kind of crucial for the thing we've been talking about.

If this is a direction we're happy with, then this part should be ready to go - still needs a renaming pass, a UI file, and tests though.

danieldresser commented:
```

johnhaddon commented on include/GafferScene/CollectPrimitiveVariables.h:
```diff
@@ -0,0 +1,92 @@
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERSCENE_COLLECTSCENES_H
```
GAFFERSCENE_COLLECTPRIMITIVEVARIABLES_H

johnhaddon commented on src/GafferScene/CollectPrimitiveVariables.cpp:
```diff
@@ -0,0 +1,259 @@
+namespace
+{
+
+static InternedString g_collectPrimitiveVariablesSuffix( "collectPrimitiveVariables:suffix" );
```
CollectScenes and CollectImages both expose this as a StringPlug. I think we should probably do the same here, and if so, "collect:primitiveVariableSuffix" seems to better match the existing names.

johnhaddon commented on src/GafferScene/CollectPrimitiveVariables.cpp:
```diff
@@ -0,0 +1,259 @@
+	if(
+		input == inPlug()->objectPlug() ||
+		input == suffixesPlug() ||
+		input == targetVariablesPlug()
```
Looking below, it appears that `targetVariablesPlug()` doesn't affect `isVariationPlug()`. These days I'm tending to avoid dirtying multiple plugs in a single block, so it's easier to match the logic in the hash/compute methods. Longer term I'm thinking of having `affects<ThePlug>()` methods to enforce this separation of logic...

johnhaddon commented on src/GafferScene/CollectPrimitiveVariables.cpp:
```diff
@@ -0,0 +1,259 @@
+	{
+		if( !( filterValue( context ) & IECore::PathMatcher::ExactMatch ) )
+		{
+			h = IECore::MurmurHash();
```
What's the intention here? A hash always needs to [start with a TypeId](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/ComputeNode.cpp#L57), so I think the right thing to do is just return, using the basic hash computed by the call to `SceneElementProcessor::hash()`.

danieldresser-ie replied:
Ah, OK.  I wasn't sure what the right thing to do here is.  At this point, I know that the value is always going to be false, so it seemed like it would make sense to return a fixed hash, but yeah, the convention makes sense.

johnhaddon commented on src/GafferScene/CollectPrimitiveVariables.cpp:
```diff
@@ -0,0 +1,259 @@
+				if( collectObjectHash != inObjectHash )
+				{
+					IECore::ConstObjectPtr collectObject = inPlug()->objectPlug()->getValue( &collectObjectHash );
+					if( collectObject != inObject )
```
Did you mean to compare pointers here, rather than pointees?

danieldresser-ie replied:
Yep, oops.

johnhaddon commented on src/GafferScene/CollectPrimitiveVariables.cpp:
```diff
@@ -0,0 +1,259 @@
+void CollectPrimitiveVariables::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	h.append( targetVariablesPlug()->hash() );
+	h.append( suffixesPlug()->hash() );
```
No need to hash this one, since we're acting on the result of `suffixesPlug()->getValue()` anyway...

danieldresser-ie replied:
We do actually need to hash it in, due a tricky thing that I originally missed:  if the upstream network isn't varying with the context variable, we don't get any hash difference from the loop, but we still use the suffix names to name the new prim vars.

johnhaddon replied:
Of course - thanks for the explanation.

johnhaddon commented on src/GafferScene/CollectPrimitiveVariables.cpp:
```diff
@@ -0,0 +1,259 @@
+	ConstStringVectorDataPtr suffixesData = suffixesPlug()->getValue();
+	const vector<string> &suffixes = suffixesData->readable();
+
+	for( std::string suffix : suffixes )
```
`const &`, so we don't copy the string?

johnhaddon commented on src/GafferScene/CollectPrimitiveVariables.cpp:
```diff
@@ -0,0 +1,259 @@
+		ConstStringVectorDataPtr suffixesData = suffixesPlug()->getValue();
+		const vector<string> &suffixes = suffixesData->readable();
+
+		for( vector<string>::const_iterator it = suffixes.begin(), eIt = suffixes.end(); it != eIt; ++it )
```
I think we're settling on using range-for for readability wherever we can.

danieldresser-ie commented:
OK, those comments look straightforward to address, the only tricky bit is "hasVariation".

> I do find the isVariation thing to be a little bit oddball. Actually, I'm not even sure I know how it's working - it depends on scene:path doesn't it, but is also being used by the SceneSwitch when computing globals/sets?

Nope, I'm being naughty and using a regular Switch.  I actually find this a bit clearer when dealing with a sequence nodes and want to disable all of them, but I guess I might be able to individually hook up all their disables.  Either way, this part of the hackery doesn't require anything new added to Gaffer ( though maybe you are going to disallow running Switch on scenes at some point ).

>  My most conservative suggestion would be to rename it to hasVariation and hide it from the NodeEditor. As an alternative though, would doing things the other way around work?
> Add a requiresVariation input plug, which when on, only applies the collected primitive variables if they have changed from the original.
> Use a (currently fictional) PrimitiveVariableExists node to drive OSLObject.enabled.
> If that works for your use case, I feel like it might be a better approach...

Hmm, intuitively feels a little bit slower, due to needing to actually evaluate the object, rather than just comparing hashes, but thinking it through, once taking caching into account, this is probably fine.  I'll have a go at this.

johnhaddon commented:
> Nope, I'm being naughty and using a regular Switch

I'm still not sure I understand how that can work. Let's say I pull on the globals downstream, so `scene:path` isn't in the context. The Switch evaluates `isVariation`, which evaluates the input object. That should barf shouldn't it?

> though maybe you are going to disallow running Switch on scenes at some point

The opposite actually. The plan is to remove SceneSwitch and ImageSwitch, but move their context-based optimisations into Switch itself. It's approaching the top of my ticket list, so should happen soon (async SceneView permitting)...

danieldresser-ie commented:
> The Switch evaluates isVariation, which evaluates the input object. That should barf shouldn't it?

Oh, hmmm, yeah, that is a bit weird that this isn't barfing.  I guess I should be able to wire up all the enables instead.

danieldresser commented:
It was too warm to go to sleep, so I took a whack at finishing this off.  I think this addresses all feedback.

Only contentious bit I can think of is that I named the suffix context variable name plug "suffixContextVariable".  "suffixVariable" would have been more consistent with the other Collect* nodes, but it felt confusing to have it right beside "targetVariables" when they are talking about completely different kinds of variables.

danieldresser-ie commented:
Closing in favour of #2703
